File: sdk/typescript/test_nomos_sdk.ts
import assert from "node:assert/strict";
import http from "node:http";

import {
  GuardedFunction,
  NomosClient,
  createActionRequest,
  guardFunction,
  guardHttpTool,
} from "./nomos_sdk.ts";
import type { ActionRequest } from "./nomos_sdk.ts";

type CapturedRequest = {
  path: string;
  headers: http.IncomingHttpHeaders;
  payload: Record<string, unknown>;
};

async function startServer(responseFactory: () => { statusCode?: number; body: Record<string, unknown> }) {
  const captured: CapturedRequest[] = [];
  const server = http.createServer((req, res) => {
    const chunks: Buffer[] = [];
    req.on("data", (chunk) => chunks.push(Buffer.from(chunk)));
    req.on("end", () => {
      const payload = JSON.parse(Buffer.concat(chunks).toString("utf8"));
      captured.push({ path: req.url ?? "", headers: req.headers, payload });
      const response = responseFactory();
      res.writeHead(response.statusCode ?? 200, { "content-type": "application/json" });
      res.end(JSON.stringify(response.body));
    });
  });

  await new Promise<void>((resolve) => server.listen(0, "127.0.0.1", () => resolve()));
  const address = server.address();
  if (!address || typeof address === "string") {
    throw new Error("failed to resolve test server address");
  }
  return {
    server,
    baseUrl: `http://127.0.0.1:${address.port}`,
    captured,
  };
}

function createClient(baseUrl: string): NomosClient {
  return new NomosClient({
    baseUrl,
    bearerToken: "dev-api-key",
    agentId: "demo-agent",
    agentSecret: "demo-agent-secret",
  });
}

async function closeServer(server: http.Server): Promise<void> {
  await new Promise((resolve) => server.close(resolve));
}

async function testAllow(): Promise<void> {
  const { server, baseUrl } = await startServer(() => ({ body: { decision: "ALLOW", execution_mode: "external_authorized" } }));
  try {
    const executed: string[] = [];
    const guarded = customTestTool<{ orderId: string }, string>({
      client: createClient(baseUrl),
      resource: (input) => `url://shop.example.com/refunds/${input.orderId}`,
      params: () => ({ method: "POST" }),
      execute: async (input) => {
        executed.push(input.orderId);
        return "refund-submitted";
      },
    });

    const result = await guarded.invoke({ orderId: "ORD-1001" });
    assert.equal(result.executed, true);
    assert.equal(result.value, "refund-submitted");
    assert.deepEqual(executed, ["ORD-1001"]);
  } finally {
    await closeServer(server);
  }
}

async function testDeny(): Promise<void> {
  const { server, baseUrl } = await startServer(() => ({ body: { decision: "DENY", reason: "deny_by_rule" } }));
  try {
    const executed: string[] = [];
    const guarded = customTestTool<{ orderId: string }, string>({
      client: createClient(baseUrl),
      resource: (input) => `url://shop.example.com/refunds/${input.orderId}`,
      params: () => ({ method: "POST" }),
      execute: async (input) => {
        executed.push(input.orderId);
        return "refund-submitted";
      },
    });

    const result = await guarded.invoke({ orderId: "ORD-1001" });
    assert.equal(result.executed, false);
    assert.equal(result.decisionResponse.decision, "DENY");
    assert.deepEqual(executed, []);
  } finally {
    await closeServer(server);
  }
}

async function testApproval(): Promise<void> {
  const { server, baseUrl } = await startServer(() => ({
    body: {
      decision: "REQUIRE_APPROVAL",
      approval_id: "apr_123",
      approval_fingerprint: "fp_123",
    },
  }));
  try {
    const executed: string[] = [];
    const guarded = customTestTool<{ orderId: string }, string>({
      client: createClient(baseUrl),
      resource: (input) => `url://shop.example.com/refunds/${input.orderId}`,
      params: () => ({ method: "POST" }),
      execute: async (input) => {
        executed.push(input.orderId);
        return "refund-submitted";
      },
    });

    const result = await guarded.invoke({ orderId: "ORD-1001" });
    assert.equal(result.executed, false);
    assert.equal(result.decisionResponse.decision, "REQUIRE_APPROVAL");
    assert.equal(result.decisionResponse.approval_id, "apr_123");
    assert.deepEqual(executed, []);
  } finally {
    await closeServer(server);
  }
}

async function testFailClosed(): Promise<void> {
  const guarded = guardFunction<string, string>({
    client: createClient("http://127.0.0.1:9"),
    buildRequest: () => createActionRequest("tool.read", "file://workspace/README.md", {}),
    execute: () => "content",
  });

  await assert.rejects(() => guarded.invoke("ignored"));
}

async function testTracePropagation(): Promise<void> {
  const { server, baseUrl, captured } = await startServer(() => ({ body: { decision: "ALLOW", execution_mode: "external_authorized", trace_id: "trace-explicit-123" } }));
  try {
    const guarded = new GuardedFunction<string, string>(
      createClient(baseUrl),
      () =>
        ({
          ...createActionRequest("tool.http", "url://shop.example.com/refunds/ORD-1001", { method: "POST" }),
          trace_id: "trace-explicit-123",
        }) as ActionRequest,
      () => "ok",
    );

    const result = await guarded.invoke("ignored");
    assert.equal(result.executed, true);
    assert.equal(captured[0]?.payload.trace_id, "trace-explicit-123");
  } finally {
    await closeServer(server);
  }
}

async function testExternalReport(): Promise<void> {
  const { server, baseUrl, captured } = await startServer(() => ({ body: { recorded: true, trace_id: "trace-custom-report", action_id: "act-custom-report", outcome: "SUCCEEDED" } }));
  try {
    const response = await createClient(baseUrl).reportExternalOutcome({
      schema_version: "v1",
      action_id: "act-custom-report",
      trace_id: "trace-custom-report",
      action_type: "payments.refund",
      resource: "payment://shop.example.com/orders/ORD-1001",
      outcome: "SUCCEEDED",
    });
    assert.equal(captured[0]?.path, "/actions/report");
    assert.equal(response.recorded, true);
  } finally {
    await closeServer(server);
  }
}

async function main(): Promise<void> {
  await testLocalGuardSafety();
  await testAllow();
  await testDeny();
  await testApproval();
  await testFailClosed();
  await testTracePropagation();
  await testExternalReport();
  console.log("TypeScript SDK tests passed");
}

await main();

async function testLocalGuardSafety(): Promise<void> {
  for (const [actionType, mode, requests] of [
    ["net.http_request", "external_authorized", 0],
    ["process.exec", "external_authorized", 0],
    ["email.send", "", 1],
    ["email.send", "nomos_executed", 1],
  ] as const) {
    const { server, baseUrl, captured } = await startServer(() => ({ body: { decision: "ALLOW", execution_mode: mode } }));
    try {
      let executed = false;
      const guard = guardFunction({
        client: createClient(baseUrl),
        buildRequest: () => createActionRequest(actionType, "inbox://local/messages/1", {}),
        execute: () => { executed = true; },
      });
      await assert.rejects(() => guard.invoke({}));
      assert.equal(executed, false);
      assert.equal(captured.length, requests);
    } finally {
      await closeServer(server);
    }
  }
}

function customTestTool<Input, Output>(config: {
  client: NomosClient;
  resource: (input: Input) => string;
  params: (input: Input) => Record<string, unknown>;
  execute: (input: Input) => Promise<Output> | Output;
}) {
  return guardFunction({
    client: config.client,
    buildRequest: (input: Input) => createActionRequest("tool.http", config.resource(input), config.params(input)),
    execute: config.execute,
  });
}

File: sdk/python/test_nomos_sdk.py
from __future__ import annotations

import json
import threading
import unittest
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

from nomos_sdk import (
    ActionRequest,
    NomosClient,
    guard_callable,
)


def custom_test_tool(*, client, resource_fn, params_fn, execute_fn):
    return guard_callable(client=client, build_request=lambda p: ActionRequest(
        "payments.refund", resource_fn(p), params_fn(p)), execute=execute_fn)


class _Handler(BaseHTTPRequestHandler):
    response_status = 200
    response_body: dict[str, object] = {"decision": "ALLOW"}
    captured_requests: list[dict[str, object]] = []

    def do_POST(self) -> None:  # noqa: N802
        length = int(self.headers.get("Content-Length", "0"))
        body = self.rfile.read(length).decode("utf-8")
        payload = json.loads(body)
        _Handler.captured_requests.append(
            {
                "path": self.path,
                "headers": dict(self.headers),
                "payload": payload,
            }
        )
        self.send_response(_Handler.response_status)
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(json.dumps(_Handler.response_body).encode("utf-8"))

    def log_message(self, format: str, *args: object) -> None:
        return


class NomosPythonWrapperTests(unittest.TestCase):
    @classmethod
    def setUpClass(cls) -> None:
        cls.server = ThreadingHTTPServer(("127.0.0.1", 0), _Handler)
        cls.thread = threading.Thread(target=cls.server.serve_forever, daemon=True)
        cls.thread.start()
        cls.base_url = f"http://127.0.0.1:{cls.server.server_address[1]}"

    @classmethod
    def tearDownClass(cls) -> None:
        cls.server.shutdown()
        cls.server.server_close()

    def setUp(self) -> None:
        _Handler.response_status = 200
        _Handler.response_body = {"decision": "ALLOW", "execution_mode": "external_authorized"}
        _Handler.captured_requests = []
        self.client = NomosClient(
            base_url=self.base_url,
            bearer_token="dev-api-key",
            agent_id="demo-agent",
            agent_secret="demo-agent-secret",
        )

    def test_custom_tool_allow_executes(self) -> None:
        executed: list[str] = []
        tool = custom_test_tool(
            client=self.client,
            resource_fn=lambda payload: f"url://shop.example.com/refunds/{payload['order_id']}",
            params_fn=lambda payload: {"method": "POST"},
            execute_fn=lambda payload: executed.append(payload["order_id"]) or "refund-submitted",
        )

        result = tool.invoke({"order_id": "ORD-1001"})

        self.assertTrue(result.executed)
        self.assertEqual(result.value, "refund-submitted")
        self.assertEqual(executed, ["ORD-1001"])

    def test_custom_tool_deny_does_not_execute(self) -> None:
        _Handler.response_body = {"decision": "DENY", "reason": "deny_by_rule"}
        executed: list[str] = []
        tool = custom_test_tool(
            client=self.client,
            resource_fn=lambda payload: f"url://shop.example.com/refunds/{payload['order_id']}",
            params_fn=lambda payload: {"method": "POST"},
            execute_fn=lambda payload: executed.append(payload["order_id"]) or "refund-submitted",
        )

        result = tool.invoke({"order_id": "ORD-1001"})

        self.assertFalse(result.executed)
        self.assertTrue(result.is_denied())
        self.assertEqual(executed, [])

    def test_custom_tool_requires_approval_does_not_execute(self) -> None:
        _Handler.response_body = {
            "decision": "REQUIRE_APPROVAL",
            "approval_id": "apr_123",
            "approval_fingerprint": "fp_123",
        }
        executed: list[str] = []
        tool = custom_test_tool(
            client=self.client,
            resource_fn=lambda payload: f"url://shop.example.com/refunds/{payload['order_id']}",
            params_fn=lambda payload: {"method": "POST"},
            execute_fn=lambda payload: executed.append(payload["order_id"]) or "refund-submitted",
        )

        result = tool.invoke({"order_id": "ORD-1001"})

        self.assertFalse(result.executed)
        self.assertTrue(result.requires_approval())
        self.assertEqual(result.decision_response["approval_id"], "apr_123")
        self.assertEqual(executed, [])

    def test_guard_callable_fails_closed_on_transport_error(self) -> None:
        broken_client = NomosClient(
            base_url="http://127.0.0.1:9",
            bearer_token="dev-api-key",
            agent_id="demo-agent",
            agent_secret="demo-agent-secret",
            timeout=0.2,
        )
        executed: list[str] = []
        guarded = guard_callable(
            client=broken_client,
            build_request=lambda value: ActionRequest(
                action_type="notes.read",
                resource="file://workspace/README.md",
                params={},
            ),
            execute=lambda value: executed.append("ran") or "content",
        )

        with self.assertRaises(Exception):
            guarded.invoke("ignored")
        self.assertEqual(executed, [])

    def test_guard_callable_propagates_explicit_trace_id(self) -> None:
        guarded = guard_callable(
            client=self.client,
            build_request=lambda value: ActionRequest(
                action_type="payments.refund",
                resource="url://shop.example.com/refunds/ORD-1001",
                params={"method": "POST"},
                trace_id="trace-explicit-123",
            ),
            execute=lambda value: "ok",
        )

        result = guarded.invoke("ignored")

        self.assertTrue(result.executed)
        self.assertEqual(_Handler.captured_requests[-1]["payload"]["trace_id"], "trace-explicit-123")

    def test_report_external_outcome(self) -> None:
        response = self.client.report_external_outcome(
            {
                "action_id": "act-custom-report",
                "trace_id": "trace-custom-report",
                "action_type": "payments.refund",
                "resource": "payment://shop.example.com/orders/ORD-1001",
                "outcome": "SUCCEEDED",
            }
        )

        self.assertEqual(_Handler.captured_requests[-1]["path"], "/actions/report")
        self.assertEqual(_Handler.captured_requests[-1]["payload"]["schema_version"], "v1")
        self.assertEqual(response["decision"], "ALLOW")


if __name__ == "__main__":
    unittest.main()
